Add instructions for use

Bottle-Water opened the pull request:
Updated the readme to include general instructions for usage of MerC, as well as specific examples if testing on bc.

SilverMight committed:
Add function to generate syntactic translations of macros

SilverMight committed:
Script for running Maki on multiple commands and getting output files

SilverMight committed:
Modularize analyze_transformations, add emit_translations to emit translations for directory of maki output files

SilverMight committed:
Multithread and fix paths

SilverMight committed:
Add logic to read in Maki out and translate files

SilverMight committed:
Handle ICE invocations, ignore duplicate defintions

We need to translation object-like macros used where ICEs are required
to anonymous enums, as const variables are not ICEs.

In addition, currently ignore macros that have multiple definitions.

SilverMight committed:
Refactor I/O of Maki analysis to one, deduped file

Previously we made multiple Maki analysis files to match the source
directory structure, however this is unnecessary. Combine everything
into one, deduplicated file (most important for compiler builtins).

In addition, emit_translations now takes in a single Maki file and no
longer a directory of results (flag: "maki_results_dir" -> "maki_results_path")

Smaller changes in run_maki_on_compile_commands.py are a new flag for
the number of threads to us and a fix to make sure we're in the correct
directory while running each compile command.

SilverMight committed:
Add flags to Maki to ignore builtin macros

SilverMight committed:
Properly handle translations of object-like macros
1) Add new IsICERepresentableByInt32 property from Maki to Invocation class.
2) Translation fixes:
    * If any invocations require an ICE and are not representable by an int, skip translation.
    * If all invocations that require ICE are representable by int, translate to enum.
    * Otherwise translate to const static variable of same type.

SilverMight committed:
Implement global scope check from Maki
1) Add IsDefinedAtGlobalScope property to Macro class.
2) Add check to return false is we are not at global scope in
   interface_equivalent predicate.
3) Simplify macro construction

SilverMight committed:
Cleanup unused variables, imports, functions
1) Remove only function and InvocationPredicate as they're unused
2) Optimize imports
3) Fix whitespace

SilverMight committed:
Remove main from analyze_transforms & reformatting

SilverMight committed:
Add logging and fix arg parse bug
Replace print statements with logger statements, which we have more
flexibility over output levels with. Users can change level to info with
new -v/--verbose flag.

SilverMight committed:
Refactor running Maki to use ProcessPoolExecutor
1) Use a ProcessPoolExecutor to run Maki.
2) Have run_maki_on_compile_command return json instead of putting it
   into a queue, and remove the queue.
   * No need to order anything and better expresses intent and trims
     down function signature
3) Trim some unnecessary variables from function signature

SilverMight committed:
Clean up arg names and add help descriptions

SilverMight committed:
Fix types and adjust exception handling
1) Remove case where we exited after capturing any exception. Printing
   the traceback would be more helpful.
2) Change return types for generate_macro_translations to match that
   there could be no translation (str | None).
3) Change MacroMap to be a dict instead of a default dict (we only
   iterate over what's placed in it)
4) Change incorrect json type to dict[str, Any]
5) Improve logging in run maki on compile commands and call .decode() on
   process stderr/stdout.

SilverMight committed:
Add missing import and remove unused

Bottle-Water committed:
bug fixes on fstring formatting and clang call

SilverMight committed:
Fix filtering of macros with multiple definitions
Logic was incorrect before and did not properly filter macros that show
up more than once.
Also abstract this into a function and add some comments explaining why
we do this.

Bottle-Water committed:
add instruction for MerC

Bottle-Water committed:
Update README.md format

Bottle-Water committed:
Update README.md format

Added code formatting to all file names

PappasBrent commented on README.md:
````diff
@@ -8,39 +8,55 @@
+The format for running the script is as follows:
+
+```
+python3 run_maki_on_compile_commands.py -p <path to maki plugin> -i <target program source directory> -c <path to compile_commands.json> -o <path to output maki analysis file, default is analysis.maki> -j <number of threads, default is number of CPUs on system> -v <verbose>
````
You may want to break this snippet up across multiple lines to make it easier to read

```suggestion
python3 run_maki_on_compile_commands.py \
    -p <path to maki plugin> \
    -i <target program source directory> \
    -c <path to compile_commands.json> \
    -o <path to output maki analysis file, default is analysis.maki> \
    -j <number of threads, default is number of CPUs on system> \
    -v <verbose>
```

PappasBrent commented on README.md:
````diff
@@ -8,39 +8,55 @@
-  `python3 run_maki_on_compile_commands.py -p ../maki/build/lib/libmaki.so -i ../bc-1.07/ -c ../bc-1.07/compile_commands.json`
+For example, running on bc may look like this: 
+```
+python3 run_maki_on_compile_commands.py -p ../maki/build/lib/libmaki.so -i ../bc-1.07/ -c ../bc-1.07/compile_commands.json
````
```suggestion
python3 run_maki_on_compile_commands.py \
    -p ../maki/build/lib/libmaki.so \
    -i ../bc-1.07/ \
    -c ../bc-1.07/compile_commands.json
```

PappasBrent commented on README.md:
````diff
@@ -8,39 +8,55 @@
-  `mkdir output`
-  `python3 emit_translations.py -i ../bc-1.07/ -m analysis.maki -o ./output`
+```
+python3 emit_translations.py -i <path to target program source directory> -m <path to maki analysis file> -o <(required) path to output directory for translation> -v <verbose>`
````
```suggestion
python3 emit_translations.py \
    -i <path to target program source directory> \
    -m <path to maki analysis file> \
    -o <(required) path to output directory for translation> \
    -v <verbose>
```

PappasBrent commented on README.md:
```diff
@@ -1,7 +1,46 @@
-# babel_fish
+# MerC
 
 # A Python tool for detecting and translating easy to translate C macros
```
Drop the leading # since this text is not part of the document title

```suggestion
A Python tool for detecting and translating easy to translate C macros
```

Bottle-Water committed:
Apply suggestions from code review

Formatting of example commands made clearer, dropped text style of description

Co-authored-by: Brent Pappas <pappasbrent@gmail.com>

Bottle-Water committed:
change clarity of dependencies

made clear that dependencies in the example are for bc

Bottle-Water committed:
remove backslash from code block

PappasBrent commented on README.md:
Won't you need these backslashes for the command to run correctly on the terminal? I think the flags won't be passed to the python script otherwise.

Bottle-Water replied:
right, totally forgot, I'll fix it now

Bottle-Water committed:
format python commands

added backslashes to python commands so they can be directly pasted into terminal and run

PappasBrent reviewed (approved):
LGTM!

Bottle-Water committed:
remove bear on config and as dependency

Bottle-Water committed:
add bear to ./configure command

You must run bear on all build commands in order to capture the compiler. See https://github.com/rizsotto/Bear/wiki/Troubleshooting
